pkg/scheduler/plugins/coescheduling/core/core.go

```diff
@@ -0,0 +1,453 @@
+	gangId, gang := pgMgr.GetGangIdAndGang(pod)
+	if gangId == "" {
+		return 0, PodGroupNotSpecified
+	}
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return 0, PodGroupNotFound
+	}
+	// first add pod to the gang's WaitingPodsMap
+	gang.addAssumedPod(pod)
+
+	gangGroup := gang.getGangGroup()
+	allGangGroupSatisfied := true
+	// only the gang itself
+	if len(gangGroup) == 0 {
+		allGangGroupSatisfied = gang.isGangResourceSatisfied()
+	} else {
+		// check each gang group
+		for _, groupName := range gangGroup {
+			gangTmp := pgMgr.cache.getGangFromCache(groupName)
+			if gangTmp != nil {
+				if !gangTmp.isGangResourceSatisfied() {
+					allGangGroupSatisfied = false
+					break
+				} else {
+					// if we cannot find the gang's group-gang  from the gangCache,
+					// we assumed that this gang should wait until timeout,
+					// user should check if the group-gang exists or the spelling mistakes
+					allGangGroupSatisfied = false
+					break
+				}
+			}
+		}
+	}
+	if !allGangGroupSatisfied {
+		return gang.WaitTime, Wait
+	}
+	return 0, Success
+}
+
+// Unreserve
+// if gang is resourceSatisfied, we only delAssumedPod
+// if gang is not resourceSatisfied and is in StrictMode, we release all the assumed pods
+func (pgMgr *PodGroupManager) Unreserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string, handle framework.Handle, pluginName string) {
+	gangId, gang := pgMgr.GetGangIdAndGang(pod)
+	if gangId == "" {
+		return
+	}
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName: %v, podName: %v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return
+	}
+	// first delete the pod from gang's waitingFroBindChildren map
+	gang.delAssumedPod(pod)
+
+	if !gang.isGangResourceSatisfied() && gang.getGangMode() == extension.GangModeStrict {
+		// release resource of all assumed children of the gang
+		handle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+			if GetNamespaceSplicingName(waitingPod.GetPod().Namespace,
+				GetGangNameByPod(waitingPod.GetPod())) == gangId {
+				klog.Errorf("unReserve rejects the pod from Gang due to timeout,gang: %v, pod: %v",
+					gangId, GetNamespaceSplicingName(pod.Namespace, pod.Name))
+				waitingPod.Reject(pluginName, "optimistic rejection in unReserve due to timeout")
+			}
+		})
+	}
+
+}
+
+// PostBind updates a PodGroup's status.
+func (pgMgr *PodGroupManager) PostBind(ctx context.Context, pod *corev1.Pod, nodeName string) {
+	gangId, gang := pgMgr.GetGangIdAndGang(pod)
+	if gangId == "" {
+		return
+	}
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName: %v, podName: %v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return
+	}
+
+	// first update CRD
+	pgFullName, pg := pgMgr.GetPodGroup(pod)
+	if pgFullName == "" || pg == nil {
+		return
+	}
+	pgCopy := pg.DeepCopy()
+	pgCopy.Status.Scheduled++
```

Comment:
if you create pg by annotation, but watched onCreate very late,  even later than first pod assigned, so "pgCopy.Status.Scheduled++" is not safe.  you should make pgCopy.Status.Scheduled = gang.GetBoundPodNum()